src/persist/src/indexed/encoding.rs

```diff
@@ -200,34 +200,108 @@ pub enum BlobTraceUpdates {
+        /// Diffs.
+        diffs: Int64Array,
+    },
 }
 
 impl BlobTraceUpdates {
     /// The number of updates.
     pub fn len(&self) -> usize {
-        self.records().len()
+        match self {
+            BlobTraceUpdates::Row(c) => c.len(),
+            BlobTraceUpdates::Both(c, _structured) => c.len(),
+            BlobTraceUpdates::Structured { timestamps, .. } => timestamps.len(),
+        }
+    }
+
+    /// The updates' timestamps as an integer array.
+    pub fn timestamps(&self) -> &Int64Array {
+        match self {
+            BlobTraceUpdates::Row(c) => c.timestamps(),
+            BlobTraceUpdates::Both(c, _structured) => c.timestamps(),
+            BlobTraceUpdates::Structured { timestamps, .. } => timestamps,
+        }
+    }
+
+    /// The updates' diffs as an integer array.
+    pub fn diffs(&self) -> &Int64Array {
+        match self {
+            BlobTraceUpdates::Row(c) => c.diffs(),
+            BlobTraceUpdates::Both(c, _structured) => c.diffs(),
+            BlobTraceUpdates::Structured { diffs, .. } => diffs,
+        }
     }
 
     /// Return the [`ColumnarRecords`] of the blob.
-    pub fn records(&self) -> &ColumnarRecords {
+    pub fn records(&self) -> Option<&ColumnarRecords> {
         match self {
-            BlobTraceUpdates::Row(c) => c,
-            BlobTraceUpdates::Both(c, _structured) => c,
+            BlobTraceUpdates::Row(c) => Some(c),
+            BlobTraceUpdates::Both(c, _structured) => Some(c),
+            BlobTraceUpdates::Structured { .. } => None,
         }
     }
 
     /// Return the [`ColumnarRecordsStructuredExt`] of the blob.
     pub fn structured(&self) -> Option<&ColumnarRecordsStructuredExt> {
         match self {
             BlobTraceUpdates::Row(_) => None,
-            BlobTraceUpdates::Both(_, structured) => Some(structured),
+            BlobTraceUpdates::Both(_, s) => Some(s),
+            BlobTraceUpdates::Structured { key_values, .. } => Some(key_values),
         }
     }
 
     /// Return the estimated memory usage of the raw data.
     pub fn goodbytes(&self) -> usize {
-        self.records().goodbytes() + self.structured().map_or(0, |e| e.goodbytes())
+        match self {
+            BlobTraceUpdates::Row(c) => c.goodbytes(),
+            // NB: we only report goodbytes for columnar records here, to avoid
+            // dual-counting the same records. (This means that our goodput % is much lower
+            // during the migration, which is an accurate reflection of reality.)
+            BlobTraceUpdates::Both(c, _) => c.goodbytes(),
+            BlobTraceUpdates::Structured {
+                key_values,
+                timestamps,
+                diffs,
+            } => {
+                key_values.goodbytes()
+                    + timestamps.values().to_byte_slice().len()
+                    + diffs.values().to_byte_slice().len()
+            }
+        }
+    }
+
+    /// Return the [`ColumnarRecordsStructuredExt`] of the blob.
```

Comment:
This comment seems incorrect? We're returning the Codec data here?

Reply:
Oops, copy-paste error - will update.